src/tree/updater_gpu_hist.cu

```diff
@@ -1308,6 +1357,37 @@ class GPUHistMaker : public TreeUpdater {
   GPUDistribution dist_;
 };
 
+class GPUHistMaker : public TreeUpdater {
```

Comment:
I think there are two choices here, one is adding a parameter `single_precision_histogram` and choose implementation at runtime, as done in this PR.  The cost is adding this wrapper class.

Another option is make it a separated updater, as done by something similar to:
``` c++
XGBOOST_REGISTER_TREE_UPDATER(GPUHistMaker<GradientPair>, "single_precision_grow_gpu_hist")

XGBOOST_REGISTER_TREE_UPDATER(GPUHistMaker<GradientPairPrecise>, "grow_gpu_hist")
```
This way those `override` modifiers can be preserved and with one less wrapper, also, a separated parameter class won't be necessary.  It makes sense to me since these two classes can indeed be considered as different updaters.

I don't mind your choice, but do you think it's worth considering?

Reply:
I thought about this but I think it is more natural for a user to toggle a training parameter than to change the updater type. I don't want users to be changing the "updater" if possible, for example there may forget to add "prune" and then get a different behaviour from tree_method="gpu_hist".